src/Razor/src/Microsoft.AspNetCore.Razor.LanguageServer/Diagnostics/RazorPullDiagnosticsEndpoint.cs

```diff
@@ -0,0 +1,90 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the MIT license. See License.txt in the project root for license information.
+
+using System;
+using System.Collections.Generic;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.AspNetCore.Razor.Language;
+using Microsoft.AspNetCore.Razor.LanguageServer.Common;
+using Microsoft.AspNetCore.Razor.LanguageServer.EndpointContracts;
+using Microsoft.AspNetCore.Razor.LanguageServer.Protocol;
+using Microsoft.VisualStudio.LanguageServer.Protocol;
+
+namespace Microsoft.AspNetCore.Razor.LanguageServer.Diagnostics;
+
+internal class RazorPullDiagnosticsEndpoint : IRazorPullDiagnosticsEndpoint
+{
+    public bool MutatesSolutionState => false;
+
+    public RegistrationExtensionResult GetRegistration(VSInternalClientCapabilities clientCapabilities)
+    {
+        return new RegistrationExtensionResult("_vs_supportsDiagnosticRequests", options: clientCapabilities.SupportsDiagnosticRequests);
+    }
+
+    public TextDocumentIdentifier GetTextDocumentIdentifier(VSInternalDocumentDiagnosticsParams request)
+    {
+        if (request.TextDocument is null)
+        {
+            throw new NotImplementedException("Why would document be null?");
+        }
+
+        return request.TextDocument;
+    }
+
+    public async Task<IEnumerable<VSInternalDiagnosticReport>?> HandleRequestAsync(VSInternalDocumentDiagnosticsParams request, RazorRequestContext context, CancellationToken cancellationToken)
+    {
+        var result = await GetPullDiagnosticsAsync(request, context, cancellationToken);
+
+        return result;
+    }
+
+    private async Task<IEnumerable<VSInternalDiagnosticReport>?> GetPullDiagnosticsAsync(VSInternalDocumentDiagnosticsParams request, RazorRequestContext requestContext, CancellationToken cancellationToken)
+    {
+        var clientNotifier = requestContext.GetRequiredService<ClientNotifierServiceBase>();
+        var documentContext = requestContext.GetRequiredDocumentContext();
+        var versionedTextDocumentIdentifier = new VersionedTextDocumentIdentifier
+        {
+            Uri = request.TextDocument!.Uri,
+            Version = documentContext.Version,
+        };
+
+        var delegatedParams = new DelegatedDiagnosticParams(versionedTextDocumentIdentifier);
+
+        var result = await clientNotifier.SendRequestAsync<DelegatedDiagnosticParams, VSInternalDiagnosticReport[]?>(
+            RazorLanguageServerCustomMessageTargets.RazorPullDiagnosticEndpointName,
+            delegatedParams,
+            cancellationToken);
+
+        var mappingService = requestContext.GetRequiredService<RazorDocumentMappingService>();
```

Comment:
Is there an advantage to this style, over just having a constructor that takes in the services needed?

Reply:
The idea is to make the handler itself stateless where possible, that way if we need to make any of our services non-singletons we don't have to change a bunch of code. It also delays paying any construction cost until the endpoint is actually used, instead of when the endpoint is created (they are all created at server-startup as part of registration), which Roslyn has found to be a big benefit for their startup time.

Those are all speculative reasoning though, if the rest of the team has strong feelings about doing this stuff in the constructor I don't feel strongly or have hard data, so it's whatever.

Reply:
Thanks for the explanation, though to make sure I understand, is "construction cost" just assigning the fields?

I don't have a strong opinion either way really, I guess I'll just reiterate my [other comment](https://github.com/dotnet/razor-tooling/pull/6949#discussion_r988383110) about getting that feedback back into the system. Should we log an issue to roll this patten out to the other single server handlers? Or do you just want to do that as part of this PR? Unless it involves jumping through a large number of hoops, I think we should aim for consistency, as much as reasonable.

Reply:
The construction cost that Roslyn was referring to is doing all the constructor calls and initialization involved in building services which are registered as a type instead of a singleton instance. This was a hard requirement, so I assume that they have found that it makes a significant difference for their scenarios.

That said, on researching my reply I've found that Ms.Ext.DI recommends [against un-necessary use of this pattern](https://learn.microsoft.com/en-us/dotnet/core/extensions/dependency-injection-guidelines#recommendations) (item 5). So lets avoid it where possible. I'll change this to use a constructor and file a follow-up to clean up unnecessary usages I may have added in the CLaSP move.

Reply:
https://github.com/dotnet/razor-tooling/issues/6955